contracts/src/v0.8/dev/ocr2/OCR2BaseUpgradeable.sol

```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../ConfirmedOwnerUpgradeable.sol";
+import "./OCR2Abstract.sol";
+import "../vendor/@openzeppelin/contracts-upgradeable/v4.8.1/contracts/proxy/utils/Initializable.sol";
+
+/**
+ * @notice Onchain verification of reports from the offchain reporting protocol
+ * @dev THIS CONTRACT HAS NOT GONE THROUGH ANY SECURITY REVIEW. DO NOT USE IN PROD.
+ * @dev For details on its operation, see the offchain reporting protocol design
+ * doc, which refers to this contract as simply the "contract".
+ * @dev This contract is meant to aid rapid development of new applications based on OCR2.
+ * However, for actual production contracts, it is expected that most of the logic of this contract
+ * will be folded directly into the application contract. Inheritance prevents us from doing lots
+ * of juicy storage layout optimizations, leading to a substantial increase in gas cost.
+ */
+abstract contract OCR2BaseUpgradeable is Initializable, ConfirmedOwnerUpgradeable, OCR2Abstract {
+  error ReportInvalid();
+
+  bool internal i_uniqueReports;
+
+  /**
+   * @dev Initializes the contract.
+   */
+  function __OCR2Base_initialize(bool uniqueReports) internal onlyInitializing {
```

Comment:
Note:
same contract as `OCR2Base`, but extends `Initializable` and swaps `constructor` for `initialize` function